File: packages/nerdster_common/lib/trust_pipeline.dart
import 'package:oneofus_common/channel_factory.dart';
import 'package:oneofus_common/statement_source.dart';
import 'package:oneofus_common/trust_statement.dart';
import 'package:oneofus_common/keys.dart';
import 'package:nerdster_common/trust_graph.dart';
import 'package:nerdster_common/trust_logic.dart';

class TrustPipeline {
  final StatementSource<TrustStatement> source;
  // If provided, keys are grouped by endpoint domain and fetched from the
  // appropriate registered channel rather than always using [source].
  final ChannelFactory? channelFactory;
  final int maxDegrees;
  final PathRequirement? pathRequirement;

  TrustPipeline(
    this.source, {
    this.channelFactory,
    this.maxDegrees = 6,
    this.pathRequirement,
  });

  static int defaultPathRequirement(int distance) {
    if (distance <= 2) return 1;
    if (distance <= 4) return 2;
    return 3;
  }

  static int permissivePathRequirement(int distance) => 1;

  static int strictPathRequirement(int distance) {
    if (distance <= 1) return 1;
    if (distance <= 3) return 2;
    return 3;
  }

  Future<TrustGraph> build(IdentityKey povIdentity) async {
    TrustGraph graph = TrustGraph(pov: povIdentity);
    Set<IdentityKey> frontier = {povIdentity};
    Set<IdentityKey> visited = {};
    Map<IdentityKey, List<TrustStatement>> statementsByIssuer = {};
    final sw = Stopwatch()..start();

    for (int depth = 0; depth < maxDegrees; depth++) {
      if (frontier.isEmpty) break;

      final keysToFetch = frontier.difference(visited)
          .where((k) => !graph.equivalent2canonical.containsKey(k))
          .toList();
      if (keysToFetch.isEmpty) break;

      final fetchMap = {for (var k in keysToFetch) k.value: null};

      Map<String, List<TrustStatement>> newStatementsMap;
      if (channelFactory != null) {
        newStatementsMap = {};
        final byUrl = <String, Map<String, String?>>{};
        for (final key in keysToFetch) {
          final url = FedKey.find(key)?.endpoint['url'] as String? ?? kNativeUrl;
          byUrl.putIfAbsent(url, () => {})[key.value] = null;
        }
        for (final entry in byUrl.entries) {
          final ch = channelFactory!.getChannel<TrustStatement>(entry.key, 'statements');
          print('[TrustPipeline] depth=$depth url=${entry.key} keys=${entry.value.length}');
          final results = await ch.fetch(entry.value);
          newStatementsMap.addAll(results);
        }
      } else {
        newStatementsMap = await source.fetch(fetchMap);
      }
      visited.addAll(keysToFetch);

      for (var entry in newStatementsMap.entries) {
        statementsByIssuer[IdentityKey(entry.key)] = entry.value;
      }

      final pr = pathRequirement ?? defaultPathRequirement;
      graph = reduceTrustGraph(
        graph,
        statementsByIssuer,
        pathRequirement: pr,
        maxDegrees: maxDegrees,
      );

      frontier = graph.distances.keys.toSet().difference(visited);
    }

    sw.stop();
    print('[TrustPipeline] done: ${sw.elapsedMilliseconds}ms  keys=${graph.distances.length}');
    return graph;
  }
}

File: functions/trust_pipeline.js
/**
 * trust_pipeline.js — JavaScript port of trust_pipeline.dart
 *
 * Orchestrates fetch + reduce cycles to build a trust graph.
 *
 * If sourceFor(url) is provided, keys discovered with a foreign endpoint
 * (via the 'with.endpoint' field in trust statements) are fetched from
 * the appropriate URL rather than the default source.
 */

const { reduceTrustGraph, defaultPathRequirement, kDefaultMaxDegrees } = require('./trust_logic');

class TrustPipeline {
  constructor(source, { maxDegrees = kDefaultMaxDegrees, pathRequirement, sourceFor } = {}) {
    this.source = source;
    this.maxDegrees = maxDegrees;
    this.pathRequirement = pathRequirement || defaultPathRequirement;
    this.sourceFor = sourceFor || null; // (url: string) => source — optional
  }

  async build(povToken, { fedRegistry = new Map(), oouCache } = {}) {
    if (!oouCache) throw new Error('TrustPipeline.build: oouCache is required');
    const visited = new Set();
    let frontier = new Set([povToken]);
    let graph = { pov: povToken, distances: new Map([[povToken, 0]]), equivalent2canonical: new Map() };

    for (let depth = 0; depth < this.maxDegrees; depth++) {
      if (frontier.size === 0) break;

      const keysToFetch = [...frontier].filter(k => !visited.has(k) && !graph.equivalent2canonical.has(k));
      if (keysToFetch.length === 0) break;

      let newStatementsMap;
      if (this.sourceFor) {
        newStatementsMap = {};
        const byUrl = new Map();
        for (const k of keysToFetch) {
          const url = fedRegistry.get(k) ?? null;
          if (!byUrl.has(url)) byUrl.set(url, []);
          byUrl.get(url).push(k);
        }
        for (const [url, keys] of byUrl) {
          const src = url ? this.sourceFor(url) : this.source;
          const fetched = await src.fetch(Object.fromEntries(keys.map(k => [k, null])));
          Object.assign(newStatementsMap, fetched);
        }
      } else {
        const fetchMap = Object.fromEntries(keysToFetch.map(k => [k, null]));
        newStatementsMap = await this.source.fetch(fetchMap);
      }

      for (const k of keysToFetch) visited.add(k);
      for (const [token, statements] of Object.entries(newStatementsMap)) {
        oouCache.set(token, statements);
      }

      graph = await reduceTrustGraph(povToken, oouCache, {
        pathRequirement: this.pathRequirement,
        maxDegrees: this.maxDegrees,
        fedRegistry,
      });

      frontier = new Set([...graph.distances.keys()].filter(k => !visited.has(k)));
    }

    return graph;
  }
}

module.exports = { TrustPipeline };
